File: python/qa_gen_viterbi_fi.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# 
# Copyright 2014 Jan Krämer.
# 
# This is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3, or (at your option)
# any later version.
# 
# This software is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this software; see the file COPYING.  If not, write to
# the Free Software Foundation, Inc., 51 Franklin Street,
# Boston, MA 02110-1301, USA.
# 

from gnuradio import gr, gr_unittest
from gnuradio import blocks, digital, analog, trellis
import celec_swig as celec
import numpy


fsm_args = {"awgn1o2_8": (2, 8, 4,
        (0, 4, 0, 4, 1, 5, 1, 5, 2, 6, 2, 6, 3, 7, 3, 7),
        (0, 3, 3, 0, 1, 2, 2, 1, 3, 0, 0, 3, 2, 1, 1, 2),
        ), 
      "awgn1o2_16": ( 2,16,4,
        (0, 8, 0, 8, 1, 9, 1, 9, 2, 10, 2, 10, 3, 11, 3, 11, 4, 12, 4, 12, 5, 13, 5, 13, 6, 14, 6, 14, 7, 15, 7, 15),
        (0, 3, 3, 0, 1, 2, 2, 1, 1, 2, 2, 1, 0, 3, 3, 0, 2, 1, 1, 2, 3, 0, 0, 3, 3, 0, 0, 3, 2, 1, 1, 2)),
        }  

class qa_gen_viterbi_fi (gr_unittest.TestCase):

    def setUp (self):
        self.tb = gr.top_block ()

    def tearDown (self):
        self.tb = None

    def test_001_t (self):
        # Trellislength
        K = 2044
        fsm = fsm_args["awgn1o2_16"]

        os = numpy.array(fsm[4], dtype=int) 
        # Setup dummy data
        #data = [1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0]
        #data = numpy.random.randint(0,2,K)
        #termination = numpy.zeros(4)
        #data = numpy.concatenate((data, termination),1)

        throt = blocks.throttle(1, 32000)
        #data = numpy.array([1,1,0,1,1,0,1,1])
        sym_table = digital.constellation_qpsk()
        #print sym_table.points()
        # Setup blocks
        #data_src = blocks.vector_source_s(map(int, data))
        data_src = blocks.file_source(1, "SG.jpeg", 0)
        #src_head = blocks.head(gr.sizeof_short*1, K)
        # TX Sim
        p2up = blocks.packed_to_unpacked_bb(1, 0)
        encoder = trellis.encoder_bb(trellis.fsm(*fsm), 0)
        modulator = digital.chunks_to_symbols_bc(sym_table.points(), 1)

        # Decoder
        viterbi_cel = celec.gen_viterbi_fi(2, 5, 9317*8, 0, -1, 
                                           sym_table.points(), os)

        # Sinks
        up2p = blocks.unpacked_to_packed_bb(1, 0)
        tx_sink = blocks.vector_sink_s(1)
        rx_sink = blocks.vector_sink_b(1)
        file_sink = blocks.file_sink(1, "decoded_SG.jpeg")
        # Connections
        self.tb.connect(data_src, throt, p2up, encoder)
        self.tb.connect(encoder, modulator)
        self.tb.connect(modulator, viterbi_cel)
        self.tb.connect(viterbi_cel, up2p, file_sink)
        #self.tb.connect(src_head, tx_sink);
        self.tb.run ()
        
        # # Check data
        #rx_output = numpy.array(rx_sink.data())
        #for k in range(0, K):
        #    self.assertEqual(int(rx_output[k]), int(data[k]))

if __name__ == '__main__':
    gr_unittest.run(qa_gen_viterbi_fi, "qa_gen_viterbi_fi.xml")

File: lib/qa_gen_viterbi_fi.cc
/* -*- c++ -*- */
/* 
 * Copyright 2014 <+YOU OR YOUR COMPANY+>.
 * 
 * This is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3, or (at your option)
 * any later version.
 * 
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this software; see the file COPYING.  If not, write to
 * the Free Software Foundation, Inc., 51 Franklin Street,
 * Boston, MA 02110-1301, USA.
 */


#include <gnuradio/attributes.h>
#include <cppunit/TestAssert.h>
#include "qa_gen_viterbi_fi.h"
#include <celec/gen_viterbi_fi.h>
#include <stdio.h>
#include <volk_fec/volk_fec.h>
#include <gnuradio/random.h>
#include <stdlib.h>

#define INF 1e9

namespace gr {
  namespace celec {

    void
    qa_gen_viterbi_fi::t1()
    {
      int vlen = 3072;  // 3072 Infowords
      int iters = 1;
      int S = 16;  // 16 States
      int O = 4;  // 4 Output Symbols
      
      double time_volk, time_generic;
      
      size_t align = volk_fec_get_alignment();
      float *alpha_generic;
      alpha_generic = (float*)volk_fec_malloc(sizeof(float)*2*S, align);
      
      float *alpha_sse;
      alpha_sse = (float*)volk_fec_malloc(sizeof(float)*2*S, align);

      for(int i = 0; i < 2*S; i++)
      {  
        alpha_generic[i] = INF;
        alpha_sse[i] = INF;
      }  
      alpha_sse[0]     = 0;
      alpha_generic[0] = 0;

      unsigned int *trace_generic;
      trace_generic = (unsigned int*)volk_fec_malloc(sizeof(int)*vlen*S, align);
      
      unsigned int *trace_sse;
      trace_sse = (unsigned int*)volk_fec_malloc(sizeof(int)*vlen*S, align);
      
      int OS[] = {0, 3, 3, 0, 1, 2, 2, 1, 1, 2, 2, 1, 0, 3, 3, 0, 2, 1, 1, 2, 3, 0, 0, 3, 3, 0, 0, 3, 2, 1, 1, 2}; 
      
      float *in;
      in = (float*) volk_fec_malloc(sizeof(float)*vlen*O, align);
      for(int i = 0; i < vlen*O; i++)
      {  
        // Create random samples ranging from -4 to 4
        in[i] = (float) (((float)::random()/RANDOM_MAX)*5);
      }
      printf("\nRunning Test: volk_fec_32f_x2_s32f_x3_32i_x3_viterbi_metric_32i_32f ...\n");
      clock_t t_volk, t;
      for(int i = 0; i < iters; i++)
      {
        t_volk = clock();
        volk_fec_32f_x2_s32f_32i_viterbi_metric_32i_manual(&trace_sse[0], &alpha_sse[0], &in[0], vlen,
                                          &OS[0],  S, "a_sse4");

        t_volk = clock() - t_volk;
        time_volk += ((float) t_volk);
  
        t = clock();
       
        volk_fec_32f_x2_s32f_32i_viterbi_metric_32i_manual(&trace_generic[0], &alpha_generic[0], &in[0], vlen,
                                                 &OS[0], S, "generic");
      
        t = clock() - t;
        time_generic += ((float) t);
        for(int i = 0; i < vlen*S; i++)
        {
          //printf("trace gen[%d]: %d trace sse[%d]: %d\n", i, trace_generic[i],i, trace_sse[i]);
          CPPUNIT_ASSERT_DOUBLES_EQUAL((double)trace_generic[i], (double)trace_sse[i],
          0.1);
          // CPPUNIT_ASSERT_DOUBLES_EQUAL(alpha_generic[i], alpha_sse[i],
          // 0.1);
        }

      }
      printf("Time for SSE: %f\n", time_volk/CLOCKS_PER_SEC/iters);
      printf("Time for generic: %f\n", time_generic/CLOCKS_PER_SEC/iters);
    }

  } /* namespace celec */
} /* namespace gr */

